File: 09/09.rs
use std::collections::HashSet;
use std::fs;
use std::time::Instant;

fn main() {
    let start = Instant::now();

    let instr: Vec<(char, usize)> = fs::read_to_string("09/input.txt")
        .expect("File not found")
        .lines()
        .map(|line| {
            (
                line.chars().next().unwrap(),
                line[2..line.len()].parse::<usize>().unwrap(),
            )
        })
        .collect();

    let (p1, p2) = part12(&instr, 9);
    println!("part1: {}", p1);
    println!("part2: {}", p2);

    println!("time: {:?}", start.elapsed());
}

fn part12(instructions: &[(char, usize)], n_knots: usize) -> (usize, usize) {
    let mut visited1: HashSet<(i16, i16)> = HashSet::new();
    let mut visited2: HashSet<(i16, i16)> = HashSet::new();

    let mut knots: Vec<(i16, i16)> = vec![(0, 0); n_knots + 1];
    visited1.insert(knots[n_knots]);
    visited2.insert(knots[n_knots]);

    for &(direction, amount) in instructions.iter() {
        for _ in 0..amount {
            match direction {
                'U' => knots[0].1 += 1,
                'D' => knots[0].1 -= 1,
                'L' => knots[0].0 -= 1,
                'R' => knots[0].0 += 1,
                _ => unreachable!(),
            }

            // drag knots behind
            for i in 1..=n_knots {
                let diff_x = knots[i - 1].0 - knots[i].0;
                let diff_y = knots[i - 1].1 - knots[i].1;

                if diff_x.abs() > 1 || diff_y.abs() > 1 {
                    knots[i].0 += diff_x.signum();
                    knots[i].1 += diff_y.signum();
                } else {
                    // "early stopping"
                    break;
                }
            }
            visited1.insert(knots[1]);
            visited2.insert(knots[n_knots]);
        }
    }
    (visited1.len(), visited2.len())
}

File: 09/09.py
from timing_util import print_elapsed, timestamp_nano


def part12(instructions, n_knots: int = 9) -> tuple[int, int]:
    def sign(x: int) -> int:
        if x > 0:
            return 1
        if x < 0:
            return -1
        return 0

    # +1 for head
    knots = [(0, 0)] * (n_knots + 1)
    visited1 = set([(knots[0])])
    visited2 = set([(knots[0])])

    for direction, amount in instructions:
        for _ in range(amount):
            head_x, head_y = knots[0]
            match direction:
                case 'U':
                    knots[0] = head_x, head_y + 1
                case 'D':
                    knots[0] = head_x, head_y - 1
                case 'L':
                    knots[0] = head_x - 1, head_y
                case 'R':
                    knots[0] = head_x + 1, head_y

            # drag knots behind
            for i in range(1, n_knots + 1):
                diff_x, diff_y = (a - b for a, b in zip(knots[i - 1], knots[i]))

                if abs(diff_x) > 1 or abs(diff_y) > 1:
                    knots[i] = tuple(x + sign(d) for x, d in zip(knots[i], (diff_x, diff_y)))
                else:
                    break

            visited1.add(knots[1])
            visited2.add(knots[-1])
        
    return len(visited1), len(visited2)

if __name__ == '__main__':
    start = timestamp_nano()

    with open('09/input.txt') as in_file:
        data = [(line[0], int(line.removesuffix('\n')[2:])) for line in in_file]

    p1, p2 = part12(data)
    print(f'part1: {p1}')
    print(f'part2: {p2}')

    print_elapsed(start)
